rfc/2022/2022-09-15_Design-System-Team_Migrate-vets-website-off-of-formation.md

```diff
@@ -0,0 +1,106 @@
+    - This will bloat the app's bundle size, but other pages and apps won't be affected
+
+We will be starting with migrating the utility classes, and the css library will export a `utilities.css` file as part of its build which can be imported in either location. All of the steps outlined here will begin with a single app, and then expanded to other apps once the pilot app has been successfully migrated.
+
+#### Remove old VADS utility classes from markup
+We will first identify a team and/or app and work with them on migrating their markup to use the new USWDS3-based utilities provided by `css-library`. As part of this we should also see if we can eliminate any unused CSS from the application's SASS file.
+
+As we eliminate `vads-u-*` classes from app code, add an [ESLint override](https://eslint.org/docs/latest/user-guide/configuring/configuration-files#how-do-overrides-work) to their app directory to throw an error if the older style of utility class is added back into a file.
+
+#### Remove component/module CSS
+
+This is related to the prerequisite for migrating to web components. For example, [this bit of app code](https://github.com/department-of-veterans-affairs/vets-website/blob/4885e8c532f77801712a6d39c0625f8ceb19a556/src/applications/vaos/new-appointment/components/VAFacilityPage/FacilitiesNotShown.jsx#L81) relies on [this CSS from formation](https://github.com/department-of-veterans-affairs/veteran-facing-services-tools/blob/f0e1d666503ecf4aafcb421bbc47fc7f76abec4a/packages/formation/sass/modules/_m-additional-info.scss#L1-L4), and that module is [included in the sitewide styles for `vets-website`](https://github.com/department-of-veterans-affairs/vets-website/blob/4885e8c532f77801712a6d39c0625f8ceb19a556/src/platform/site-wide/sass/style.scss#L16). As we get to a point where we can be sure that certaion SASS modules are no longer being used, we can remove them from `vets-website`, giving us a good idea of how far along we are in the migration while also reducing the bundle size.
+
+#### Try to replace "Custom CSS" with utility classes
+
+Many applications have custom CSS [like this](https://github.com/department-of-veterans-affairs/vets-website/blob/fad1a404cc2c7a8c33153d159523c14888d6fa03/src/applications/gi/sass/partials/_gi-search-page.scss#L303-L310) where they are making declarations in a CSS file when there are existing CSS utility classes for the exact same declarations that could be used instead. We will also want to do an audit to identify places like this where utility classes can be used and then make changes to use the appropriate utilities.
+
+This will have a performance benefit as well, since an audit of that particular CSS class would reveal that the `search-result-tag` CSS class isn't used in application code at all, so we are shipping unused CSS that should be deleted. As we begin to rely more on utility classes and developers write less CSS, this type of thing will happen less often.
```

Comment:
These are distinct concerns. We could audit for unused CSS right now and remove it as a separate step. In fact, any app team could do that today. Worth mentioning but I wouldn't tie this directly to the "custom CSS" argument. We should be doing this either via audit or via automation.

Reply:
I'm not sure what you mean with this comment. I'm suggesting that JSX & SASS files be audited for CSS classes. For example, there could be jsx like:

```jsx

<div className="my-app-some-component">{foo}</div>

```

And the app's SASS file could have:

```css

.my-app-some-component {
    margin: 0;
    background-color: white;
}
```

The JSX could then be migrated to:

```jsx
<div className="vads-margin-0 vads-bg-white">{foo}</div>
```

and the CSS class removed. The audit to discover migration opportunities like this could also reveal CSS that is simply no longer used, and removing it could be a performance improvement in the form of a smaller file size.

Reply:
It's the same point and discussion as above. Taking your example: "Every page is getting CSS meant for the dashboard because that CSS exists in the site-wide SASS file." That is a bug. That team should not have put that code there. That team could put app-specific CSS in a file that only that application loads or, to your point, use the CSS class library. But I'm not convinced one is more performant than the other site-wide. If a team chooses to maintain a CSS file that only loads in their application and they prefer to do that over using the CSS library then I don't think we should dictate to them that they must switch to using the CSS library. I do think we could aggressively move their CSS out of a shared file and into an app specific file that only their application loads (and gzip it).